File: llm_proxy/target.py
"""Target URL parsing and path joining."""

from __future__ import annotations

from typing import TypedDict
from urllib.parse import urlsplit

from .constants import DEFAULT_PORTS


class ParsedTargetUrl(TypedDict):
    scheme: str
    host: str
    port: int
    base_path: str
    display_url: str


def parse_target_url(raw_target_url: str) -> ParsedTargetUrl:
    parsed = urlsplit(raw_target_url)
    if parsed.scheme not in DEFAULT_PORTS or not parsed.hostname:
        raise ValueError("target_url must look like http://host[:port][/base-path] or https://host[:port][/base-path].")
    return {
        "scheme": parsed.scheme,
        "host": parsed.hostname,
        "port": parsed.port or DEFAULT_PORTS[parsed.scheme],
        "base_path": parsed.path.rstrip("/"),
        "display_url": raw_target_url.rstrip("/"),
    }


def join_target_path(base_path: str, request_path: str) -> str:
    """Join an upstream base path with a client request path."""
    if not base_path:
        return request_path
    if not request_path.startswith("/"):
        request_path = f"/{request_path}"
    if (
        request_path == base_path
        or request_path.startswith(f"{base_path}/")
        or request_path.startswith(f"{base_path}?")
    ):
        return request_path
    return f"{base_path}{request_path}"

File: src/proxy/target.ts
const DEFAULT_PORTS = {
  http: 80,
  https: 443,
} as const;

export interface ParsedTargetUrl {
  readonly scheme: keyof typeof DEFAULT_PORTS;
  readonly host: string;
  readonly port: number;
  readonly basePath: string;
  readonly displayUrl: string;
}

export function parseTargetUrl(rawTargetUrl: string): ParsedTargetUrl {
  const schemeMatch = /^(https?):\/\//i.exec(rawTargetUrl);
  if (schemeMatch === null) {
    throw invalidTargetUrl();
  }

  let parsed: URL;
  try {
    parsed = new URL(rawTargetUrl);
  } catch {
    throw invalidTargetUrl();
  }
  const scheme = parsed.protocol.slice(0, -1).toLowerCase();
  if ((scheme !== "http" && scheme !== "https") || parsed.hostname === "") {
    throw invalidTargetUrl();
  }

  return {
    scheme,
    host: stripIpv6Brackets(parsed.hostname),
    port: parsed.port === "" ? DEFAULT_PORTS[scheme] : parsePort(parsed.port),
    basePath: parsed.pathname.replace(/\/+$/, ""),
    displayUrl: rawTargetUrl.replace(/\/+$/, ""),
  };
}

function parsePort(value: string): number {
  const port = Number(value);
  if (!Number.isInteger(port) || port < 0 || port > 65_535) {
    throw invalidTargetUrl();
  }
  return port;
}

function stripIpv6Brackets(hostname: string): string {
  return hostname.startsWith("[") && hostname.endsWith("]") ? hostname.slice(1, -1) : hostname;
}

function invalidTargetUrl(): TypeError {
  return new TypeError(
    "target_url must look like http://host[:port][/base-path] or https://host[:port][/base-path].",
  );
}
